### ISSUE TITLE
Server unable to handle ANY requests when "Http Endpoint" Activity Path contains "//"

### ISSUE THREAD

**Issue by polferov** [elsa-workflows/elsa-core#4415]
# The Issue

Disclaimer: I use path and route interchangeably, as the HttpEndpoint activity calls it path and ASP.NET Core calls it route.

When creating an HttpEndpoint activity and entering a route that contains two consecutive slashes ("//"), e.g. when accidentally pasting (or getting via js) an absolute url (e.g. "https://abc.xyz"), an exception will be thrown on ANY request to the server.

The exception(s) are

```
Microsoft.AspNetCore.Routing.Patterns.RoutePatternException: The route template separator character '/' cannot appear consecutively. It must be separated by either a parameter or a literal value.
   at Microsoft.AspNetCore.Routing.Patterns.RoutePatternParser.Parse(String pattern)
   at Microsoft.AspNetCore.Routing.Template.TemplateParser.Parse(String routeTemplate)
```

and

```
System.ArgumentException: The route template separator character '/' cannot appear consecutively. It must be separated by either a parameter or a literal value. (Parameter 'routeTemplate')
 ---> Microsoft.AspNetCore.Routing.Patterns.RoutePatternException: The route template separator character '/' cannot appear consecutively. It must be separated by either a parameter or a literal value.
   at Microsoft.AspNetCore.Routing.Patterns.RoutePatternParser.Parse(String pattern)
   at Microsoft.AspNetCore.Routing.Template.TemplateParser.Parse(String routeTemplate)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Routing.Template.TemplateParser.Parse(String routeTemplate)
   at Elsa.Activities.Http.Services.RouteMatcher.Match(String routeTemplate, String requestPath)
   at Elsa.Activities.Http.Middleware.HttpEndpointMiddleware.<>c__DisplayClass2_0.<InvokeAsync>b__0(String route)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at Elsa.Activities.Http.Middleware.HttpEndpointMiddleware.InvokeAsync(HttpContext httpContext, IOptions`1 options, IWorkflowLaunchpad workflowLaunchpad, IWorkflowInstanceStore workflowInstanceStore, IWorkflowRegistry workflowRegistry, IWorkflowBlueprintReflector workflowBlueprintReflector, IRouteMatcher routeMatcher, ITenantAccessor tenantAccessor, IEnumerable`1 contentParsers)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

# Proposed Solution

As far as I'm aware, there is currently no option for input validation.
Therefore another solutoion must be found:

Within the RouteTableExtensions class, filter for the issue causing double slashes:

```diff
    public static void AddRoutes(this IRouteTable routeTable, IEnumerable<Trigger> triggers)
    {
-        var paths = Filter(triggers).Select(Deserialize).Select(x => x.Path).ToList();
+        var paths = Filter(triggers).Select(Deserialize).Select(x => x.Path)
+            .Where(x => !x.Contains("//")).ToList();

        routeTable.AddRange(paths);
    }

    public static void AddRoutes(this IRouteTable routeTable, IEnumerable<Bookmark> bookmarks)
    {
-        var paths = Filter(bookmarks).Select(Deserialize).Select(x => x.Path).ToList();
+        var paths = Filter(bookmarks).Select(Deserialize).Select(x => x.Path)
+            .Where(x => !x.Contains("//")).ToList();

        routeTable.AddRange(paths);
    }
```

This way, no routes with "//" will be added to the route table, therefore therefore, nothing will be triggering the exception.

Aditionally, within the HttpEndpoint class, add a guard to the OnExecute method. E.g.

```diff
-        protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context) => context.WorkflowExecutionContext.IsFirstPass ? ExecuteInternal(context) : Suspend();
+        protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
+        {
+            if(Path.Contains("//"))
+                throw new Exception("Path cannot contain double slashes (//)");
+            return context.WorkflowExecutionContext.IsFirstPass ? ExecuteInternal(context) : Suspend();
+        }

```

This prevents users not knowing what the issue is, as otherwise, routes would just not be added to the route table without returning any errors. But this way, users can see that their Activity failed and read the Journal to find the cause.

---

PS: I'm a newbie to Elsa, so im very open to other suggestions and/or new input.

**Comment by sfmskywalker** [comment]
Hi @polferov , thank you for reporting the issue with great detail, much appreciated!

I agree, until we have proper input validation, the best course of action for now seems to be to make sure 
erroneous paths are filtered out of the route table to prevent other HTTP endpoints from not functioning.

**Comment by sfmskywalker** [comment]
I retract my last statement that it may not be necessary to have HTTP Endpoint throw an exception, because if a path doesn't exist in the route table (such as `//hello-world`), it will simply attempt to find a trigger or bookmark matching the request path, in which case the workflow will happily execute.

However, since we do ultimately want to allow valid paths only, I agree that for now we should throw an exception, and later on, implement input validation to prevent erroneous path from being entered in the first place.

**Comment by sfmskywalker** [comment]
Fixed via 76720103b48ed57f5130658093d68811716d5c10
